Multi-tenancy in accessing node images via Pod API

stlaz opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR makes it possible to distinguish different tenants that attempt to use node-local images in their Pods. The distinction is based on the credentials that are used to pull an image.

#### Which issue(s) this PR fixes:
Related to https://github.com/kubernetes/enhancements/issues/2535

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
New configuration is introduced to the kubelet that allows it to track container images and the list of authentication information that lead to their successful pulls . This data is persisted across reboots of the host and restarts of the kubelet.

The kubelet ensures any image requiring credential verification is always pulled if authentication information from an image pull is not yet present, thus enforcing authentication / re-authentication. This means an image pull might be attempted even in cases where a pod requests the `IfNotPresent` image pull policy, and might lead to the pod not starting if its pull policy is `Never` and is unable to present authentication information that lead to a previous successful pull of the image it is requesting.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2535-ensure-secret-pulled-images/README.md
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2535-ensure-secret-pulled-images/README.md
```

TODOs:
- [x] add feature gate handling
- [x] fix unit tests
- [x] tidy up the code
- [x] add more tests
- [x] implement cleanup on GC
- [x] implement startup cache housekeeping
- [x] use a proper encoder to get versioned config files output

Open Questions:
- [x] should we salt the credential hashes to prevent introspection of credential reuse for different repositories?
    - probably not if we keep the dirs and files accessible to kubelet only
- [x] should MustAttemptImagePull amend an `ImagePullRecord` if it permits image retrieval with a secret that **a)** matches coordinates but has a different creds hash; **b)** matches creds hash but has different secret coordinates?
    - we probably want to do this to avoid unnecessary pulls but it is odd to do it in this method

stlaz committed:
credentialprovider: track kube secrets as creds sources in DockerKeyrings

stlaz committed:
imagepullmanager: add v1alpha1 config API

stlaz committed:
make update

stlaz committed:
kubelet: implement image pull policies

stlaz committed:
kubelet: add the ImagePullManager interface

stlaz committed:
kubelet: modify KubeletConfiguration API with image pull policies

Also adds PreloadedImagesVerificationAllowlist to API exceptions list
for missing list type as this is not a part of the REST API.

stlaz committed:
kubelet imagepuller: PullImage() - accept TrackAuthConfigs directly

The image puller's PullImage() method should be just a dumb pull
without any further logic. Make it accept everything it needs
to pull an image and defer any other magic to the image manager.

stlaz committed:
implement a file-based image pull manager

stlaz committed:
kubelet: integrate the image pull manager

liggitt commented on cmd/kubelet/app/options/options.go:
```diff
@@ -420,6 +420,8 @@ func AddKubeletConfigFlags(mainfs *pflag.FlagSet, c *kubeletconfig.KubeletConfig
 
 	fs.Int32Var(&c.RegistryPullQPS, "registry-qps", c.RegistryPullQPS, "If > 0, limit registry pull QPS to this value.  If 0, unlimited.")
 	fs.Int32Var(&c.RegistryBurst, "registry-burst", c.RegistryBurst, "Maximum size of a bursty pulls, temporarily allows pulls to burst to this number, while still not exceeding registry-qps. Only used if --registry-qps > 0")
+	fs.StringVar(&c.ImagePullCredentialsVerificationPolicy, "image-pull-credentials-verification-policy", c.ImagePullCredentialsVerificationPolicy, "") // TODO:
+	fs.StringSliceVar(&c.PreloadedImagesVerificationAllowlist, "preloaded-images-verification-allowlist", c.PreloadedImagesVerificationAllowlist, "List of preloaded images that can be accessed without credential verification. Can only be used with --image-pull-credentials-verification-policy=NeverVerifyAllowlistedImages.")
```
no need to expose these as flags, new stuff can be set only in the kubelet config file

stlaz committed:
kubelet: GC for image pull managers

BenTheElder commented:
/remove-sig api-machinery
/sig architecture

fedebongio commented:
/remove-sig api-machinery

fedebongio commented:
/remove-sig api-machinery

stlaz commented:
/milestone v1.32
This should now be ready for a review

haircommander commented:
/milestone v1.32

pacoxu commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -381,11 +389,13 @@ func (im *realImageGCManager) GarbageCollect(ctx context.Context, beganGC time.T
 		amountToFree := capacity*int64(100-im.policy.LowThresholdPercent)/100 - available
 		klog.InfoS("Disk usage on image filesystem is over the high threshold, trying to free bytes down to the low threshold", "usage", usagePercent, "highThreshold", im.policy.HighThresholdPercent, "amountToFree", amountToFree, "lowThreshold", im.policy.LowThresholdPercent)
-		freed, err := im.freeSpace(ctx, amountToFree, freeTime, images)
+		usedImages, freed, err := im.freeSpace(ctx, amountToFree, freeTime, images)
```
imagesLeft？（probably not used for some cases.）

cici37 commented:
/remove-sig api-machinery

liggitt commented on pkg/kubelet/images/image_pull_policies.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+func NewNeverVerifyAllowListedPullPolicy(allowList []string) ImagePullPolicyEnforcer {
+	allowListSet := sets.New(allowList...)
```
does this handle the suffix globbing (in the allowList) and the tag/sha truncation of image described in the KEP?

stlaz replied:
I completely forgot to come back to this, I'll fix that ofc

liggitt commented on pkg/kubelet/images/image_pull_policies.go:
```diff
@@ -0,0 +1,66 @@
+	kubeletconfig1betav1 "k8s.io/kubelet/config/v1beta1"
+)
+
+type ImagePullPolicyEnforcer func(image string, imagePresent, imageRecordsExist bool) bool
```
document what imagePresent and imageRecordsExist booleans mean

liggitt commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -270,11 +268,25 @@ func NewKubeGenericRuntimeManager(
+		}
+	}
+
+	imageGCHooks = append(imageGCHooks, imagePullManager.PruneUnknownRecords)
```
this should only be added if the feature is enabled, right?

liggitt commented on staging/src/k8s.io/kubelet/config/v1beta1/types.go:
```diff
@@ -41,6 +41,16 @@ const (
 	HairpinNone = "none"
 )
 
+// TODO: document all of the below
```
document, and mirror in the internal types

liggitt commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,663 @@
+		return nil, err
+	}
+
+	m.startupHousekeeping(ctx)
```
nit: "startup" sounds like this is spawning a background goroutine ... suggest a name like `initialize()` if this is doing blocking things that will complete before this returns

liggitt commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,663 @@
+	// walk the pulling directory for any pull intent records
+	err := processDirFiles(f.pullingDir,
+		func(filePath string, fileContent []byte) error {
+			var intent imagemanagerv1alpha1.ImagePullIntent
+			if _, _, err := f.decoder.Decode(fileContent, nil, &intent); err != nil {
```
(these comments apply anywhere we're decoding data from files)

1. decode using a scheme with both the versioned and internal types registered
2. decode into an internal type
3. use the returned object (don't assume the input object is used)
4. check the returned gvk to make sure it is a valid type
5. use a scheme/codec configured to be strict (error on unrecognized fields in the file)

liggitt commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,663 @@
+}
+
+// writePulledRecord writes an ImagePulledRecord into the f.pulledDir directory.
+func (f *FileBasedImagePullManager) writePulledRecord(image, imageRef string, credentials *imagemanagerv1alpha1.ImagePullCredentials) error {
```
document what it means if credentials is nil (we call it this way in initialize and error cases)

liggitt commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,663 @@
+	return nil
+}
+
+func (f *FileBasedImagePullManager) RecordImagePulled(image, imageRef string, credentials *imagemanagerv1alpha1.ImagePullCredentials) {
```
is it an error to call this with credentials == nil? if so, document that in the interface and check here to prevent accidents

mikebrow replied:
anonymous pulled case covered here?

liggitt commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,663 @@
+	f.dereferenceImagePullIntent(image)
+}
+
+// dereferenceImagePullIntent decreses the number of how many times image pull
```
nit: decrementImagePullIntent might be clearer than dereferenceImagePullIntent?

liggitt commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,663 @@
+	if f.intentCounters[image] > 0 {
+		return
+	}
+	f.intentCounters[image] = 0 // safeguard, we should never actually go below zero
```
drop instead of accumulate zero values?

```suggestion
	delete(f.intentCounters, image)
```

liggitt commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,663 @@
+			pulledRecord.CredentialMapping[sanitizedImage] = *credentials
+		}
+	} else if err == nil {
+		if _, _, err := f.decoder.Decode(pulledFile, nil, &pulledRecord); err != nil {
```
same comments about decoding into internal types strictly and checking the GVK

liggitt commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,663 @@
+		func(filePath string, fileContent []byte) error {
+			var intent imagemanagerv1alpha1.ImagePullIntent
+			if _, _, err := f.decoder.Decode(fileContent, nil, &intent); err != nil {
+				klog.ErrorS(err, "skipping file, failed to deserialize to ImagePullIntent", "filePath", filePath)
```
should we skip or delete?

mikebrow replied:
my read is the only way this happens is kubelet reset while recording intent which should be before the pull request so.. delete?

stlaz replied:
delete seems safe here, I'll switch to it

liggitt commented on pkg/kubelet/images/image_pull_manager.go:
a general structural comment on this impl... the logical operations are *very* intermixed with the file operations (check for existence, read existing record, update/write record, delete record)

separating the file operations from the logical operations would make the implementations of each a lot smaller / simpler, and make it much more straightforward to wrap or replace the always-read-from-disk impl with a work-from-memory-and-write-through-to-disk impl in the future while leaving all the logical operations untouched

```go
ListImagePullIntents() (images []string, error)
ImagePullIntentExists(image string) (bool, error)
CreateImagePullIntent(image string) error
DeleteImagePullIntent(image string) error
ListImagePulledRecords() ([]*ImagePulledRecord, error)
GetImagePulledRecord(imageRef string) (record *ImagePulledRecord, exists bool, err error)
WriteImagePulledRecord(imageRef string, record *ImagePulledRecord) error 
DeleteImagePulledRecord(imageRef string) error
```

stlaz replied:
I definitely agree, I actually realized last week that something like this would've been super useful for future implementations.

I think what originally lead me to the more "flat" design was the extent of the work. Would it be ok to address this in a follow-up?

stlaz replied:
I'll try to work on this some more then

liggitt commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,663 @@
+
+	pulledFilePath := filepath.Join(f.pulledDir, cacheFilename(imageRef))
+
+	var imageRecordExists bool
```
imageRecordExists is confusing... does this mean this is an image the kubelet is involved in pulling?

stlaz replied:
> does this mean this is an image the kubelet is involved in pulling?

Yes, that would be the case. Would something like `pulledByKubelet` read better?

stlaz replied:
renamed to `imagePulledByKubelet`

liggitt commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,663 @@
+}
+
+// writeFile writes `content` to the file with name `filename` in directory `dir`
+func writeFile(dir, filename string, content []byte) error {
```
we have to do atomic replaces to avoid partially written data in files... that means write to a tmp file, then rename over the destination file

liggitt commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,663 @@
+
+var _ ImagePullManager = &FileBasedImagePullManager{}
+
+// NamedLockSet stores named locks in order to allow to partition context access
```
nit: I'd group unrelated content in individual files (makes the files smaller / more targeted / easier to review), e.g. 

lock.go
interface.go
noop.go


I'd also avoid exporting types that aren't used outside this package

stlaz replied:
I'd ideally like to address this along with https://github.com/kubernetes/kubernetes/pull/128152#discussion_r1826046534 in a follow-up PR.

I'd like to see if I can create a submodule here where I could split these files accordingly.

drewhagen commented:
Hey @stlaz @liggitt @pacoxu @derekwaynecarr 
I see this PR is tagged for the `v1.32` release and is being actively worked on. Awesome work!

Friendly reminder that the code freeze is starting `02:00 UTC Friday November 8th 2024` (a little less than **1 week from now**). Please help make sure this PR has both `lgtm` and `approved` labels before the code freeze. Thanks! 🚀

mikebrow commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -78,33 +97,25 @@ func NewImageManager(recorder record.EventRecorder, imageService kubecontainer.I
 // imagePullPrecheck inspects the pull policy and checks for image presence accordingly,
 // returning (imageRef, error msg, err) and logging any errors.
-func (m *imageManager) imagePullPrecheck(ctx context.Context, objRef *v1.ObjectReference, logPrefix string, pullPolicy v1.PullPolicy, spec *kubecontainer.ImageSpec, imgRef string) (imageRef string, msg string, err error) {
+func (m *imageManager) imagePullPrecheck(ctx context.Context, objRef *v1.ObjectReference, logPrefix string, pullPolicy v1.PullPolicy, spec *kubecontainer.ImageSpec, containerImage string) (imageRef string, msg string, err error) {
```
ok but it's an imgRef..

liggitt replied:
the *input* to this function is the literal `image` field value from the Pod spec, right? is that an "imgRef"? (I don't actually know what the difference between imgRef / image spec / container image is)

the returned value (resolved from the container runtime) *is* an image ref?

mikebrow replied:
ok so actually prechecking the container image not the imgRef?

stlaz replied:
based on https://github.com/kubernetes/kubernetes/pull/128152#discussion_r1827998988, I landed on `requestedImage` to stress that this is the original string that entered this module from the outside.

haircommander reviewed (changes requested):
mostly naming notes, couple of structural, but overall headed in the right direction!

mikebrow commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -143,19 +171,95 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, objRef *v1.ObjectR
 		RuntimeHandler: podRuntimeHandler,
 	}
 
-	imageRef, message, err = m.imagePullPrecheck(ctx, objRef, logPrefix, pullPolicy, &spec, imgRef)
```
if it's imgRef I know what it is.. if it's containerImage does that mean manifest digest for a platform or index or the source...:tag  or something else

mikebrow replied:
currently only used for logging the never pull not present error .. but yeah this should be container image as the ref won't help the user much

mikebrow replied:
and for consistency would be bad if someone did code thinking it was container image not ref.

mikebrow replied:
looks like he fixed it.. does matter as ref isn't readable user would have to call status for the ref..

mikebrow replied:
imgRef vs imageRef glad this is being fixed confusing

mikebrow commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -143,19 +171,95 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, objRef *v1.ObjectR
+			})
+		}
+
+		pullRequired := len(imageRef) == 0 || m.imagePullManager.MustAttemptImagePull(containerImage, imageRef, imagePullSecrets)
```
nod the old logic didn't have the precheck early exit for never pull.. so now we are checking twice..

mikebrow replied:
and two cases for never pull now one where the image is not present the other were it is but we believe we had information identifying which credentials have access?

stlaz replied:
> one where the image is not present the other where it is but we believe we had information identifying which credentials have access?

exactly, those are two different checks:
1. is the image present?
2. can I access the present image?

I was hoping I could somehow merge both the checks in the pull precheck but I couldn't figure out any sensible approach, so I'm leaving these as two separate checks.

mikebrow commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -143,19 +171,95 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, objRef *v1.ObjectR
+
+	pullCredentials, _ := keyring.Lookup(repoToPull)
+
 	if imageRef != "" {
```
needs a comment.. precheck error'd out for never pull above.. falling through to pull image at bottom of fund

stlaz replied:
I moved the `PullPolicy == Never` check outside the above `imageRef != ""` block and added an explanatory comment.

mikebrow commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -82,33 +97,25 @@ func NewImageManager(recorder record.EventRecorder, nodeKeyring credentialprovid
 // imagePullPrecheck inspects the pull policy and checks for image presence accordingly,
 // returning (imageRef, error msg, err) and logging any errors.
-func (m *imageManager) imagePullPrecheck(ctx context.Context, objRef *v1.ObjectReference, logPrefix string, pullPolicy v1.PullPolicy, spec *kubecontainer.ImageSpec, imgRef string) (imageRef string, msg string, err error) {
+func (m *imageManager) imagePullPrecheck(ctx context.Context, objRef *v1.ObjectReference, logPrefix string, pullPolicy v1.PullPolicy, spec *kubecontainer.ImageSpec, containerImage string) (imageRef string, msg string, err error) {
```
true.. imageObject maybe?

stlaz replied:
Would `imageName` work?

stlaz commented on pkg/kubelet/apis/config/v1beta1/defaults.go:
```diff
@@ -173,6 +173,9 @@ func SetDefaults_KubeletConfiguration(obj *kubeletconfigv1beta1.KubeletConfigura
 		obj.HairpinMode = kubeletconfigv1beta1.PromiscuousBridge
 	}
+	if obj.ImagePullCredentialsVerificationPolicy == "" {
+		obj.ImagePullCredentialsVerificationPolicy = kubeletconfigv1beta1.NeverVerifyPreloadedImages
```
I'll remove the defaulting. Empty value will then only be interpreted as `NeverVerifyPreloaded` if the feature is enabled.

stlaz commented on pkg/kubelet/apis/config/imagemanager/v1alpha1/register.go:
```diff
@@ -0,0 +1,48 @@
+)
+
+// GroupName is the group name used in this package
+const GroupName = "imagemanager.kubelet.config.k8s.io"
```
I chose the group based on the KEP (https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2535-ensure-secret-pulled-images/README.md#cache-directory-structure).

I'll change it to the more generic kubelet config API group, it seems to be more in line with how the rest of the config API works for the kubelet. I'll change the respective mentions in the KEP afterwards.

stlaz commented:
Thank you all for the reviews! I started addressing the comments, I'll be marking them as resolved as I go, there might be a few force pushes to this PR before all is addressed.

stlaz commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,597 @@
+
+func (f *FileBasedImagePullManager) PruneUnknownRecords(imageList []string, until time.Time) {
+	// TODO: also cleanup the lock maps for intent/pull records?
+	panic("implement me")
```
It's implemented in a later commit. The changes there were a bit more awkward so I moved it to a separate commit. I will merge the latter one in here once we agree that it is ok.

stlaz commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -19,48 +19,36 @@ package kuberuntime
 	"k8s.io/kubernetes/pkg/features"
 	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
-	"k8s.io/kubernetes/pkg/util/parsers"
+	// "k8s.io/kubernetes/pkg/util/parsers"
```
oh, that one should've been removed

stlaz commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,597 @@
+
+	hash := sha256.New()
+	hash.Write([]byte(credBytes))
+	// hash.Write([]byte(salt))
```
Hmm, it's true that whatever secret is stored on the disk would likely be readable by the kubelet anyway. Keeping the dir and files locked to kubelet should be adequate. I'll remove the comments :+1:

stlaz commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -171,30 +275,32 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, objRef *v1.ObjectR
+		image, imagePullResult.pullDuration.Truncate(time.Millisecond), imagePullDuration, imagePullResult.imageSize), klog.Info)
 	metrics.ImagePullDuration.WithLabelValues(metrics.GetImageSizeBucket(imagePullResult.imageSize)).Observe(imagePullDuration.Seconds())
 	m.backOff.GC()
+	finalPullCredentials = imagePullResult.credentialsUsed
```
Oh, I had no idea. That's a really good catch

pacoxu commented:
/retest

stlaz commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,597 @@
+		var err error
+		pulledFile, err = os.ReadFile(pulledFilePath)
+		if err == nil {
+			imageRecordExists = true
```
unfortunately, you need to handle each of the files differently - in particular, you're interested in the `PulledRecord` further on, but the `PullIntent` is more or less irrelevant.

You'd have to store the content of the files in two different variables, and at that point you're getting close to the original branching :slightly_frowning_face:

stlaz commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,597 @@
+	pullingDir string
+	pulledDir  string
+
+	imagePullPolicy ImagePullPolicyEnforcer
```
renamed to `requireCredentialVerification`

liggitt commented on pkg/kubelet/apis/config/fuzzer/fuzzer.go:
```diff
@@ -120,6 +120,7 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
 			obj.EnableSystemLogHandler = true
 			obj.MemoryThrottlingFactor = ptr.To(rand.Float64())
 			obj.LocalStorageCapacityIsolation = true
+			obj.ImagePullCredentialsVerificationPolicy = ""
```
drop this line, shouldn't be needed any more

liggitt commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -30,6 +30,7 @@ import (
 	logsapi "k8s.io/component-base/logs/api/v1"
 	"k8s.io/component-base/metrics"
 	tracingapi "k8s.io/component-base/tracing/api/v1"
+	kubeletconfigv1alpha1 "k8s.io/kubelet/config/v1alpha1"
```
switch all references to constants to kubeletconfig, drop unnecessary casts

liggitt commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -266,6 +267,30 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration, featur
+			allErrors = append(allErrors, fmt.Errorf("invalid configuration: option %q specified for imagePullCredentialsVerificationPolicy. Valid options are %q, %q, %q or %q",
+				kc.ImagePullCredentialsVerificationPolicy, kubeletconfigv1alpha1.NeverVerify, kubeletconfigv1alpha1.NeverVerifyPreloadedImages, kubeletconfigv1alpha1.NeverVerifyAllowlistedImages, kubeletconfigv1alpha1.AlwaysVerify))
+		}
+
```
need validation the values in the PreloadedImagesVerificationAllowlist field are well-formed

liggitt commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -143,19 +173,80 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, objRef *v1.ObjectR
+
+func (m *imageManager) pullImage(ctx context.Context, logPrefix string, objRef *v1.ObjectReference, podUID types.UID, image string, imgSpec kubecontainer.ImageSpec, pullCredentials []credentialprovider.TrackedAuthConfig, podSandboxConfig *runtimeapi.PodSandboxConfig) (imageRef, message string, err error) {
+	if err := m.imagePullManager.RecordPullIntent(image); err != nil {
+		return "", fmt.Sprintf("Failed to record image pull intent for container image %q: %v", image, err), err
 	}
 
-	backOffKey := fmt.Sprintf("%s_%s", pod.UID, imgRef)
+	var pullSucceeded bool
+	var finalPullCredentials *credentialprovider.TrackedAuthConfig
+	defer func() {
+		if pullSucceeded {
+			m.imagePullManager.RecordImagePulled(image, imageRef, trackedToImagePullCreds(finalPullCredentials))
+		} else {
+			m.imagePullManager.RecordImagePullFailed(image)
+		}
+	}()
```
wrap this in a feature gated block:

```diff
  var pullSucceeded bool
  var finalPullCredentials *credentialprovider.TrackedAuthConfig
+ if utilfeature.DefaultFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) {
  	if err := m.imagePullManager.RecordPullIntent(image); err != nil {
  		return "", fmt.Sprintf("Failed to record image pull intent for container image %q: %v", image, err), err
  	}
  	defer func() {
  		if pullSucceeded {
  			m.imagePullManager.RecordImagePulled(image, imageRef, trackedToImagePullCreds(finalPullCredentials))
  		} else {
  			m.imagePullManager.RecordImagePullFailed(image)
  		}
  	}()
+ }
```

liggitt commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,719 @@
+		return nil, err
+	}
+
+	if gvk.Kind != "ImagePullIntent" ||
+		gvk.Group != kubeletconfigv1alpha1.SchemeGroupVersion.Group {
+		return nil, fmt.Errorf("unknown gvk for %s: %v", "ImagePullIntent", gvk)
+	}
```
I think we can actually drop the gvk check if we're type-asserting *kubeletconfiginternal.ImagePullIntent below

liggitt commented on pkg/kubelet/images/image_pull_manager.go:
```diff
@@ -0,0 +1,719 @@
+		return nil, err
+	}
+
+	if gvk.Kind != "ImagePulledRecord" ||
+		gvk.Group != kubeletconfigv1alpha1.SchemeGroupVersion.Group {
+		return nil, fmt.Errorf("unknown gvk for %s: %v", "ImagePulledRecord", gvk)
+	}
```
I think we can actually drop the gvk check if we're type-asserting *kubeletconfiginternal.ImagePulledRecord below

liggitt reviewed (commented):
I focused on the config / type validations and feature gating, I didn't dig into the pull manager verifier logic

fedebongio commented:
/triage accepted
/remove-sig api-machinery

liggitt commented on pkg/credentialprovider/keyring.go:
```diff
@@ -50,6 +54,42 @@ type providersDockerKeyring struct {
+		return nil, fmt.Errorf("cannot construct TrackedAuthConfig with a nil AuthConfig")
+	}
+
+	hashAuthConfig, err := hashAuthConfig(c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate hash for AuthConfig: %w", err)
+	}
```
ok... skip the lazy bit, limit the hash calculation to when the feature is enabled and drop the error bit (if we can't hash, the hash stays empty and we don't get the benefit of it in comparisons later)

stlaz committed:
introduce the KubeletEnsureSecretImages featuregate

stlaz committed:
kubelet: pullmanager: move to a separate package

stlaz commented:
I got delayed a bit by some other PRs but I thinkg I finally addressed the latest comments. Also, the last two commits include the refactoring with the pluggable cache records accessor and a new directory structure.

stlaz committed:
kubelet: pullmanager: write to pulled record if secret matches during query

stlaz commented:
The last commit also adds a cache write during `MustAttemptImagePull()` if the credentials matched in order to further prevent needless pull.
Since it's just a single commit we can just remove that behavior if it's unexpected.

haircommander commented:
/retest

haircommander commented on pkg/kubelet/images/fs_pullrecords.go:
```diff
@@ -0,0 +1,275 @@
+
+// FSPullRecordsAccessor uses the filesystem to read/write ImagePullIntent/ImagePulledRecord
+// records.
+type FSPullRecordsAccessor struct {
```
should this type be public?

stlaz replied:
It might be better to make it private and only use `PullRecordsAccessor` from the inside I suppose.

haircommander commented on pkg/kubelet/kubelet_pods_test.go:
```diff
@@ -6196,7 +6196,7 @@ func TestKubelet_HandlePodCleanups(t *testing.T) {
 			kl.podWorkers = podWorkers
 			originalPodSyncer := podWorkers.podSyncer
-			syncFuncs := newPodSyncerFuncs(originalPodSyncer)
+			syncFuncs := newTestPodSyncerFuncs(originalPodSyncer)
```
hm why is this in this PR? sorry if I'm missing something but I don't see this referenced elsewhere

stlaz replied:
I just saw it on the way and it threw me off when I was searching for something. It's not important, but I felt like fixing it.

haircommander replied:
if you happen to do another push, can you drop this commit and add it to a separate PR? it is already a large PR so the fewer things we can fit for the alpha impl the better IMO

stlaz replied:
done, I also unexported the fsPullRecordsAccessor

stlaz replied:
I'll create a separate commit later, I think there might've been other things that could just live in tests

haircommander commented:
LGTM, great work!

mepi262 commented:
@stlaz 
It seems that this pull request has conflicts.
Would you resolve the conflicts?

stlaz commented:
rebased on current master

pacoxu commented:
/milestone clear

pacoxu commented:
/test pull-kubernetes-e2e-gce

pacoxu commented on test/featuregates_linter/test_data/versioned_feature_list.yaml:
```diff
@@ -642,6 +642,12 @@
+  - default: false
+    lockToDefault: false
+    preRelease: Alpha
+    version: "1.32"
```
Same here.

stlaz commented:
Bumped the FG version where appropriate and added basic e2e tests.

stlaz commented:
/retest-required
infra issues + flakes

stlaz commented:
/retest
infra issues + flakes

BenTheElder commented on test/e2e/common/node/runtime.go:
```diff
@@ -274,37 +278,13 @@ while true; do sleep 1; done
 				if hasSecret {
-					// The service account only has pull permission
-					auth := `
-{
```
same comment thread: https://github.com/kubernetes/kubernetes/pull/128372#discussion_r1918965807

this approach to testing is not very sustainable and needs an alternative (perhaps a local in-cluster registry)

stlaz replied:
Thank you for the warning. For now I'm going to remove the e2e tests per yesterday's review session we had, we agreed that the code can be thoroughly unit tested and therefore the e2e tests may not be necessary.

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -241,14 +241,18 @@ func (f *PullManager) MustAttemptImagePull(image, imageRef string, podSecrets []
 		for _, cachedSecret := range cachedCreds.KubernetesSecrets {
 
 			if podSecret.CredentialHash == cachedSecret.CredentialHash {
```
Scenario 1:
1. namespace A, secret A, credential Foo works and gets recorded with hash(Foo)
2. checking namespace B, secret B, credential Foo matches (because of the hash)
3. credential rotates to Bar, A/A and B/B secret get updated to Bar
4. checking namespace B, secret B, credential Bar would not match


Recording the newly verified secret in step 2 would let it match in step 4.

This does expose us to append the number of secrets recorded without bound in the case where namespaces or pull secrets are being deleted / recreated. We could solve that with:
1. a hard cap on the number of secrets in a pull record where we would eagerly record new already-verified secrets (like 100)
2. make credential garbage collection be informed by namespaces/pull secret names

I think the backstop limit is fine for now, maybe make a note in the KEP to think about informing pruning by namespace / pull secret name data for beta.

stlaz replied:
For now I stuck to the 100 limit. Documented this at https://github.com/kubernetes/enhancements/pull/5115

liggitt commented on pkg/kubelet/images/pullmanager/locks.go:
```diff
@@ -0,0 +1,75 @@
+func (n *namedLockSet) Unlock(name string) {
+	// cannot be locked by the global lock as it would deadlock once GlobalLock() gets activated
+	if _, ok := n.locks[name]; ok {
+		n.locks[name].Unlock()
```
this leaks lock entries from the map... need to delete them somehow

however, that makes concurrent calls to Unlock("foo") and Lock("foo") racy... if the Unlock is going to delete the entry, and Lock is counting on the entry being present

stlaz replied:
We could use lock striping - preallocate, say, 151 locks and then
```go
func (s *stripedLockSet) Lock(name string) {
    fooHash := hashKey(name) // hashKey(string) ->uint32
    s.locks[fooHash % s.locksNum].Lock()
}

func (s *stripedLockSet) GlobalLock() {
    for i := range s.locks {
        s.locks[i].Lock()
    }
}
```
We'd no longer need an additional global lock as we don't need to prevent anyone from adding new locks into the struct and it's no longer unbounded.

liggitt commented on pkg/kubelet/images/pullmanager/locks.go:
```diff
@@ -0,0 +1,75 @@
+// for the same named lock.
+func (n *namedLockSet) Unlock(name string) {
+	// cannot be locked by the global lock as it would deadlock once GlobalLock() gets activated
+	if _, ok := n.locks[name]; ok {
```
reads from `n.locks` have to be done under global lock... concurrently adding an unrelated entry to locks and reading from locks will panic

liggitt commented on pkg/kubelet/images/pullmanager/locks.go:
```diff
@@ -0,0 +1,75 @@
+	}()
+	// This call cannot be guarded by the global lock as it would block the access
+	// to the other locks
+	n.locks[name].Lock()
```
I don't see how this can be done outside a global lock once we add deletion of individual entries to the Unlock path

liggitt commented on pkg/kubelet/images/pullmanager/fs_pullrecords.go:
```diff
@@ -0,0 +1,275 @@
+	targetPath := filepath.Join(dir, filename)
+	tmpPath := targetPath + ".tmp"
+	if err := os.WriteFile(tmpPath, content, 0600); err != nil {
+		return fmt.Errorf("failed to create temporary file %q: %w", tmpPath, err)
```
do a best-effort delete of the tmp file here in case it was partially written (ignore error)

liggitt commented on pkg/kubelet/images/pullmanager/fs_pullrecords.go:
```diff
@@ -0,0 +1,275 @@
+			klog.V(4).InfoS("path is a directory, skipping", "path", path)
+			return nil
+		}
+
```
maybe skip files that don't match cacheFilename pattern (skip any .tmp files, spurious files that got created in the dir by some other process that don't match the cacheFilename pattern)?

liggitt commented on pkg/kubelet/images/pullmanager/fs_pullrecords.go:
```diff
@@ -0,0 +1,275 @@
+
+		if d.IsDir() {
+			klog.V(4).InfoS("path is a directory, skipping", "path", path)
+			return nil
```
I think we have to return SkipDir here to actually skip

liggitt commented on pkg/kubelet/images/pullmanager/fs_pullrecords.go:
```diff
@@ -0,0 +1,275 @@
+
+		fileContent, err := os.ReadFile(path)
+		if err != nil {
+			klog.ErrorS(err, "skipping file, failed to read", "filePath", path)
+			return nil
```
I think we should aggregate this into `errors`, but continue by returning `nil`

liggitt commented on pkg/kubelet/images/pullmanager/fs_pullrecords.go:
```diff
@@ -0,0 +1,275 @@
+		}
+
+		if err := fileAction(path, fileContent); err != nil {
+			return err
```
aggregate this into `errors`, but continue by returning nil

liggitt commented on pkg/kubelet/images/pullmanager/fs_pullrecords.go:
```diff
@@ -0,0 +1,275 @@
+		}
+
+		if d.IsDir() {
+			klog.V(4).InfoS("path is a directory, skipping", "path", path)
```
avoid logging this, it doesn't really matter

liggitt commented on pkg/kubelet/images/pullmanager/fs_pullrecords.go:
```diff
@@ -0,0 +1,275 @@
+func processDirFiles(dirName string, fileAction func(filePath string, fileContent []byte) error) error {
+	return filepath.WalkDir(dirName, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
```
I *think* returning err here short-circuits all file reading, which is not what we want

I think we should aggregate errors at the top level of processDirFiles, but keep walking the directory to read as many files as we can

```go
var errors []error
walkErr := filepath.WalkDir(dirName, func(path string, d fs.DirEntry, err error) error {
  if err != nil {
    errors = append(errors, err)
    return nil
  }
  ...
})

if walkErr != nil {
  errors = append(errors, walkErr)
}
return errors.NewAggregate(errors)
```

liggitt commented on pkg/kubelet/images/pullmanager/fs_pullrecords.go:
```diff
@@ -0,0 +1,275 @@
+		func(filePath string, fileContent []byte) error {
+			intent, err := decodeIntent(f.decoder, fileContent)
+			if err != nil {
+				klog.V(4).InfoS("skipping file, failed to deserialize to ImagePullIntent", "filePath", filePath, "err", err)
+				return nil
```
I *think* returning the error here is better than logging it, and processDirFiles will aggregate it (along with any read / walk errors)

liggitt replied:
same comment applies everywhere we call processDirFiles

liggitt commented on pkg/kubelet/images/pullmanager/fs_pullrecords.go:
```diff
@@ -0,0 +1,275 @@
+		return nil, false, err
+	}
+
+	pulledRecord, err := decodePulledRecord(f.decoder, recordBytes)
+	return pulledRecord, true, err
```
it's a little weird to not do the same check that `pulledRecord.ImageRef == imageRef` that we're doing in the ImagePullIntentExists function ... maybe double check that and return `nil, false, nil` in that case?

liggitt commented on pkg/kubelet/images/pullmanager/fs_pullrecords.go:
```diff
@@ -0,0 +1,275 @@
+
+func (f *fsPullRecordsAccessor) DeleteImagePullIntent(image string) error {
+	intentRecordPath := filepath.Join(f.pullingDir, cacheFilename(image))
+	if err := os.Remove(intentRecordPath); err != nil && !os.IsNotExist(err) {
+		return err
+	}
```
simplify to return nil for os.IsNotExist, err otherwise

liggitt commented on staging/src/k8s.io/kubelet/config/v1beta1/types.go:
```diff
@@ -210,6 +211,10 @@ type KubeletConfiguration struct {
 	// Default: 10
 	// +optional
 	RegistryBurst int32 `json:"registryBurst,omitempty"`
+	// +optional
+	ImagePullCredentialsVerificationPolicy v1alphav1.ImagePullCredentialsVerificationPolicy `json:"imagePullCredentialsVerificationPolicy,omitempty"`
+	// +optional
+	PreloadedImagesVerificationAllowlist []string `json:"preloadImagesVerificationAllowlist,omitempty"`
```
copy the godoc from the internal type here, this is what gets generated into config API docs

liggitt commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -155,6 +155,24 @@ type KubeletConfiguration struct {
+	//       - all images require credential reverification
+	ImagePullCredentialsVerificationPolicy string
+	// preloadedImagesVerificationAllowlist specifies a list of images that are
+	// excerpted from credential reverification for the "NeverVerifyAllowlistedImages"
```
```suggestion
	// exempted from credential reverification for the "NeverVerifyAllowlistedImages"
```

liggitt commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -155,6 +155,24 @@ type KubeletConfiguration struct {
+	// preloadedImagesVerificationAllowlist specifies a list of images that are
+	// excerpted from credential reverification for the "NeverVerifyAllowlistedImages"
+	// `imagePullCredentialsVerificationPolicy`.
+	// The list accepts a full path segment wildcard suffix "/*".
```
document that image tag and digest are not considered and should not be specified in these entries

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_policies.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+func (p *NeverVerifyAllowlistedImages) AllowPattern(pattern string) error {
+	pattern = strings.TrimSpace(pattern)
```
I would error if the input has leading / trailing spaces, rather than masking this (move into validateAllowlistImagePattern)

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_policies.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+func validateAllowlistImagePattern(pattern string) error {
+	if strings.ContainsRune(pattern[:len(pattern)-1], '*') {
```
does this panic if the pattern is ""?

liggitt replied:
simplify to avoid possible index panics, etc

```go
func getTrimmedPattern(pattern string) (string, bool, error) {
  trimmedPattern := ""
  isPrefixPattern = false
  if strings.HasSuffix(pattern, "/*") {
    trimmed = strings.TrimSuffix(pattern, "*")
    isPrefixPattern = true
  } else {
    trimmedPattern = pattern
  }

  if len(trimmedPattern) == 0 {
    return error
  }
  if strings.ContainsRune(trimmedPattern, '*') {
    return error
  }
  if !isPrefixPattern && strings.HasSuffix(trimmedPattern, "/") {
    return error
  }
  return nil
}
```

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_policies.go:
```diff
@@ -0,0 +1,151 @@
+		return err
+	}
+
+	if strings.HasSuffix(pattern, "/*") {
```
use boolean returned from getTrimmedPattern

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_policies.go:
```diff
@@ -0,0 +1,151 @@
+	return false
+}
+
+func (p *NeverVerifyAllowlistedImages) AllowPattern(pattern string) error {
```
I would probably inline this into the constructor.. AllowPattern is confusing to export and sounds like RequiresVerification

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,479 @@
+
+func (f *PullManager) RecordImagePulled(image, imageRef string, credentials *kubeletconfiginternal.ImagePullCredentials) {
+	if credentials == nil || (len(credentials.KubernetesSecrets) == 0 && !credentials.NodePodsAccessible) {
+		klog.ErrorS(
+			fmt.Errorf("credentials must not be nil, and they must contain at least one kubernetes secret coordinate or must set NodePodsAccessible=true"),
+			"attempted to write an ImagePulledRecord with invalid credentials",
+			"image", image, "imageRef", imageRef, "credentials", credentials)
+		return
```
Can't we record an image pulled record with an empty CredentialMapping in that case, rather than erroring?  It seems like we could call `writePulledRecord(image, imageRef, credentials)` with nil or empty credentials

That would record that the kubelet was involved in pulling the image, but means any future attempts to use the image will have to reverify the credentials, right?

stlaz replied:
I think I would be in favour of that rather than keeping the pull intent -> I'll make it so.

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,479 @@
+
+	pulledRecord, _, err := f.recordsAccessor.GetImagePulledRecord(imageRef)
+	if err != nil {
+		return err
```
This is an interesting case... if we can't read the existing record for some reason, _should_ we error and return early? That would mean we'd fail to write the new record, and that a corrupted pulled record file cannot self heal, and will keep us from remembering new credentials used to pull it are valid

stlaz replied:
This is indeed likely the best place to attempt self-heal. There may not be much value keeping a record we cannot automatically read.

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,479 @@
+// after any tag or digest were removed from it.
+func pulledRecordMergeNewCreds(orig *kubeletconfiginternal.ImagePulledRecord, imageNoTagDigest string, newCreds *kubeletconfiginternal.ImagePullCredentials) (*kubeletconfiginternal.ImagePulledRecord, bool) {
+	ret := orig.DeepCopy()
+	if newCreds == nil {
+		return ret, false
+	}
```
nit, move this above ret = orig.DeepCopy()?

also, return early `if newCreds == nil || (!newCreds.NodePodsAccessible && len(newCreds.KubernetesSecrets) == )`?

liggitt reviewed (commented):
did a quick pass on pkg/kubelet/images/pullmanager/image_pull_manager.go

benjaminapetersen commented on api/api-rules/violation_exceptions.list:
```diff
@@ -13,13 +13,15 @@ API rule violation: list_type_missing,k8s.io/kubelet/config/v1alpha1,CredentialP
 API rule violation: list_type_missing,k8s.io/kubelet/config/v1alpha1,CredentialProvider,Env
 API rule violation: list_type_missing,k8s.io/kubelet/config/v1alpha1,CredentialProvider,MatchImages
 API rule violation: list_type_missing,k8s.io/kubelet/config/v1alpha1,CredentialProviderConfig,Providers
+API rule violation: list_type_missing,k8s.io/kubelet/config/v1alpha1,ImagePullCredentials,KubernetesSecrets
```
Quick note to fix this. :)

pacoxu commented:
/test pull-kubernetes-node-e2e-containerd-alpha-features
/test pull-kubernetes-node-e2e-containerd-standalone-mode-all-alpha
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features

pacoxu commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features

pacoxu commented on pkg/credentialprovider/keyring.go:
```diff
@@ -72,11 +118,13 @@ type AuthConfig struct {
+		dk.creds = make(map[string][]TrackedAuthConfig)
 	}
 	for loc, ident := range cfg {
 		creds := AuthConfig{
```
Not critical.

stlaz replied:
See `NewTrackedAuthConfig()` - the requested change is actually what happens there

pacoxu commented:
/lgtm
/priority important-soon

enj commented:
/milestone v1.33

stlaz commented:
Rebased on current master. There was a conflict at `pkg/kubelet/apis/config/v1beta1/defaults_test.go`

stlaz commented:
/test pull-kubernetes-integration
unrelated failures in `TestPolicyAdmission`

pacoxu commented:
/lgtm
after rebase
/test pull-kubernetes-node-e2e-containerd-alpha-features
kindly ping @liggitt @mikebrow  @yujuhong

wendy-ha18 commented:
Hello @stlaz @pacoxu 
Thanks for actively working on the PR. Just friendly reminder that code freeze is starting at 02:00 UTC Friday 21st March 2025 (about 2 weeks from now), and while there is still time, we want to ensure that each PR has a chance to be merged on time.

Please make sure this PR has both lgtm and approved labels before the code freeze. Thanks!

stlaz commented:
Rebased on the current master.
There was a conflict in imports in `pkg/kubelet/kuberuntime/kuberuntime_manager.go`

pacoxu commented:
/lgtm
/assign @haircommander 
the requested change should be addressed.

stlaz commented:
Rebased on the current master which was now housing the latest image pulling changes from @aramase.

Most notable change:
Since this PR is moving the keyring to `pkg/kubelet/images/image_manager.go`, the new keyring additions were moved here, too.

stlaz commented:
> the requested change should be addressed.

Please see https://github.com/kubernetes/kubernetes/pull/128152#discussion_r1991029956, I believe the request was actually addressed already.

stlaz commented:
`pull-kubernetes-unit` is an unrelated client-go unit tests failure

mrunalp commented on pkg/kubelet/images/fs_pullrecords.go:
```diff
@@ -0,0 +1,301 @@
+/*
```
nit: Can we fold this change to where it is first introduced?

mrunalp commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -32,7 +32,7 @@ import (
 	"k8s.io/kubernetes/pkg/features"
 	kubeletconfig "k8s.io/kubernetes/pkg/kubelet/apis/config"
-	"k8s.io/kubernetes/pkg/kubelet/images"
+	imagepullmanager "k8s.io/kubernetes/pkg/kubelet/images/pullmanager"
```
nit: Comment for this  whole commit: Can we move it where it is first introduced?

stlaz replied:
I would like to keep this move as a separate commit if you don't mind. I feel like moving some of the bits that were previously available to all into a separate package might be worth a commit on its own and makes the flow somewhat easier to follow.

stlaz commented on pkg/kubelet/images/pullmanager/locks.go:
```diff
@@ -16,60 +16,52 @@ limitations under the License.
+
+// StripedLockSet allows context locking based on string keys, where each key
+// is mapped to a an index in a size-limited slice of locks.
+type StripedLockSet struct {
```
Absolutely, I made this commit to address a comment and was waiting to see if this solution was ok.

stlaz commented:
Rebased on current master and squashed a few commits per review

stlaz commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+			}
+		case err == nil && pulledRecord != nil:
+			imagePulledByKubelet = true
+		case err != nil && exists: // record exists but has invalid format
```
looking at this again after some time:
the error is not propagated to the below code, we don't flip `imagePulledByKubelet` to true... Is this correct?

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+			ImageRef:          imageRef,
+			CredentialMapping: make(map[string]kubeletconfiginternal.ImagePullCredentials),
+		}
+		if credentials != nil {
```
comment that the existence of the pulled record for the imageRef is sufficient to indicate this was a kubelet-pulled image, and if there's no entry for the image in the credential mapping, the kubelet fails safe and requires credential validation

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+// `imageNoTagDigest` is the content of the `image` field from a pod's container
+// after any tag or digest were removed from it.
+func pulledRecordMergeNewCreds(orig *kubeletconfiginternal.ImagePulledRecord, imageNoTagDigest string, newCreds *kubeletconfiginternal.ImagePullCredentials) (*kubeletconfiginternal.ImagePulledRecord, bool) {
+	if newCreds == nil || (!newCreds.NodePodsAccessible && len(newCreds.KubernetesSecrets) == 0) {
+		return orig, false
+	}
```
nit, split these into multiple returns for readability and commenting, e.g.

```
if newCreds == nil {
  // no new credential information to record
  return orig, false
}
if len(newCreds.KubernetesSecrets) == 0 && !newCreds.NodePodsAccessible {
  // we don't have any secret credentials or node-wide access to record
  // TODO(...): add in serviceaccount dimension check
  return orig, false
}
```

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+		if ret.CredentialMapping == nil {
+			ret.CredentialMapping = make(map[string]kubeletconfiginternal.ImagePullCredentials)
+		}
+		ret.CredentialMapping[imageNoTagDigest] = *newCreds
```
return early in this case

```
ret.LastUpdatedTime = metav1.Time{Time: time.Now()}
return ret, true
```

and unnest below

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+		}
+		ret.CredentialMapping[imageNoTagDigest] = *newCreds
+	} else {
+		if newCreds.NodePodsAccessible {
```
try to simplify this with an early check and return if the existing `selectedCreds` is already NodePodsAccessible

```
if selectedCred.NodePodsAccessible {
  // already accessible to all pods on the node
  return ret, false
}

if newCreds.NodePodsAccessible {
	selectedCreds.NodePodsAccessible = true
	selectedCreds.KubernetesSecrets = nil
	ret.LastUpdatedTime = metav1.Time{Time: time.Now()}
	return ret, true
}

// ... merge pull secrets
```

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+// imagePullSecretLess is a helper function to define ordering in a slice of
+// ImagePullSecret objects.
+func imagePullSecretLess(a, b kubeletconfiginternal.ImagePullSecret) int {
+	if a.Namespace < b.Namespace {
+		return -1
+	} else if a.Namespace > b.Namespace {
+		return 1
+	}
```
```
if i := strings.Compare(a.Namespace, b.Namespace); i != 0 {
  return i
}
if i := strings.Compare(a.Name, b.Name); i != 0 {
  return i
}
return strings.Compare(a.UID, b.UID)
```

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+// object from `new`.
+//
+// The returned slice is sorted by Namespace, Name and UID (in this order).
+func mergePullSecrets(orig, new []kubeletconfiginternal.ImagePullSecret) []kubeletconfiginternal.ImagePullSecret {
```
make this return `changed bool` as well so we don't have to deepcompare

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+	}
+
+	for _, s := range new {
+		credSet[kubeSecretCoordinates{UID: s.UID, Namespace: s.Namespace, Name: s.Name}] = s.CredentialHash
```
```
key := kubeSecretCoordinates{UID: s.UID, Namespace: s.Namespace, Name: s.Name}
existingHash, exists := credSet[key]
if !exists || existingHash != s.CredentialHash {
  changed = true
  credSet[key] = s.CredentialHash
}
```

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+	for _, s := range new {
+		credSet[kubeSecretCoordinates{UID: s.UID, Namespace: s.Namespace, Name: s.Name}] = s.CredentialHash
+	}
+
```
```
if !changed {
  return orig, false
}
```

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+		pulledRecord, exists, err = f.recordsAccessor.GetImagePulledRecord(imageRef)
+		switch {
+		case err == nil && pulledRecord == nil:
+			if exists, err := f.recordsAccessor.ImagePullIntentExists(image); err != nil {
```
think about whether we can check our refCount map as a fast positive check first... if we don't have a ref count, then we do need to fall back to this call

stlaz replied:
I like the idea. I added that in the code + commented on why we need to do both in case the refCount is 0.

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+		var err error
+		var exists bool
+		pulledRecord, exists, err = f.recordsAccessor.GetImagePulledRecord(imageRef)
+		switch {
```
let's move all the cases handling `pulledRecord != nil`, `exists == true`, `err != nil` *first*

```
case err != nil:
  return error
case exists:
  imagePulledByKubelet = true
case pulledRecord != nil:
  imagePulledByKubelet = true
default:
  if exists, err := f.recordsAccessor.ImagePullIntentExists(image); err != nil {
  }
```

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+		return true
+	}
+
+	if !f.requireCredentialVerification(image, len(imageRef) > 0, imagePulledByKubelet) {
```
drop the `len(imageRef) > 0` arg to this function, I think it is always true

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+					// While we're only matching at this point, we want to ensure this secret is considered valid in the future
+					// and so we make an additional write to the cache.
+					// writePulledRecord() is a noop in case the secret with the update hash already appears in the cache.
+					if err := f.writePulledRecord(image, imageRef, &kubeletconfiginternal.ImagePullCredentials{KubernetesSecrets: []kubeletconfiginternal.ImagePullSecret{podSecret}}); err != nil {
```
nit, rename to `writePulledRecordIfChanged` to make it clearer it handles no-oping

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+	if newCreds == nil || (!newCreds.NodePodsAccessible && len(newCreds.KubernetesSecrets) == 0) {
+		return orig, false
+	}
+	ret := orig.DeepCopy()
```
since we're calling down into here in read paths (to add new credentials in MustAttemptImagePull), it would be ideal to delay calling DeepCopy until we know we actually need to change `ret`

stlaz replied:
Did that + commented in the function doc to hopefully make sure this won't get refactored out in the future.

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+			}
+
+			if secretCoordinatesMatch {
+				if len(cachedCreds.KubernetesSecrets) < writeRecordWhileMatchingLimit { // we already know the hashes did not match, no need to check again
```
let's just make this symmetrical so if the code moves it is still correct / efficient, and drop the need for a comment

```suggestion
				if !hashesMatch && len(cachedCreds.KubernetesSecrets) < writeRecordWhileMatchingLimit {
```

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+	f.pulledAccessors.GlobalLock()
+	defer f.pulledAccessors.GlobalUnlock()
+
+	pulledRecords, err := f.recordsAccessor.ListImagePulledRecords()
```
if we get partial success here (pulledRecords have some entries, *and* some errors were encountered decoding), I think we should log AND continue to clean up what we can

liggitt commented on pkg/kubelet/images/pullmanager/fs_pullrecords.go:
```diff
@@ -0,0 +1,301 @@
+	return pulledRecord, true, err
+}
+
+func (f *fsPullRecordsAccessor) ListImagePulledRecords() ([]*kubeletconfiginternal.ImagePulledRecord, error) {
```
document that this can return partial success (non-empty records AND an error indicating issues were encountered during list)

liggitt commented on pkg/kubelet/images/pullmanager/fs_pullrecords.go:
```diff
@@ -0,0 +1,301 @@
+	return writeFile(f.pullingDir, cacheFilename(image), intentBytes.Bytes())
+}
+
+func (f *fsPullRecordsAccessor) ListImagePullIntents() ([]*kubeletconfiginternal.ImagePullIntent, error) {
```
document that this can return partial success (non-empty records AND an error indicating issues were encountered during list)

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+// This method is not thread-safe and it should only be called upon the creation
+// of the PullManager.
+func (f *PullManager) initialize(ctx context.Context) {
+	pullIntents, err := f.recordsAccessor.ListImagePullIntents()
```
same comment here, if we get partial success here (pullIntents have some entries, and some errors were encountered decoding), I think we should log AND continue to clean up what we can

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+	// Each of the images known to the CRI might consist of multiple tags and digests,
+	// which is what we track in the ImagePullIntent - we need to go through all of these
+	// for each image.
+	for _, image := range images {
```
nit: can we use a different name for `image` / `images` here (containerImage / containerImages) and save `image` for the string image spec found by searchForExistingTagDigest

stlaz replied:
I'll use `imageObj` for the results of the object listing and keep `image` for the image+tag/digest

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+	// which is what we track in the ImagePullIntent - we need to go through all of these
+	// for each image.
+	for _, image := range images {
+		imageName := searchForExistingTagDigest(inFlightPulls, image)
```
```suggestion
		image := searchForExistingTagDigest(inFlightPulls, containerImage)
```

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+// searchForExistingTagDigest loop through the `image` RepoDigests and RepoTags
+// and tries to find a digest/tag in `trackedImages`, which is a map of
+// containerImage -> pulling intent path.
+func searchForExistingTagDigest(trackedImages sets.Set[string], image kubecontainer.Image) string {
```
make it clear what the string this returns is (`image` spec or `imageRef`)

It's possible for there to be *multiple* image pull intents for differently named references to the same underlying imageRef, so this needs to return `images []string`

(also, rename trackedImages to inFlightPulls)

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+			continue
+		}
+
+		if err := f.recordsAccessor.DeleteImagePullIntent(imageName); err != nil {
```
loop over all the image names for this container image / imageRef, and delete their intents

stlaz commented on pkg/kubelet/images/pullmanager/image_pull_policies.go:
```diff
@@ -0,0 +1,161 @@
+		return trimmedPattern, false, nil
+	}
+
+	if len(trimmedPattern) == 1 {
+		return "", false, fmt.Errorf("at least registry hostname is required")
+	}
```
looks like this was protecting us from just `/*`,

I'll move it under a `if wildCard` block

stlaz commented on pkg/credentialprovider/plugin/plugins.go:
```diff
@@ -87,11 +87,12 @@ func NewExternalCredentialProviderDockerKeyring(podNamespace, podName, podUID, s
 
 	for _, p := range k.providers {
-		keyring.Add(p.Provide(image))
+		// TODO: modify the credentialprovider.CredentialSource to contain the SA/pod information
```
created https://github.com/kubernetes/kubernetes/issues/130808 for this purpose

stlaz commented on pkg/kubelet/images/pullmanager/fs_pullrecords.go:
```diff
@@ -0,0 +1,301 @@
+// a successful write, it move the temp file in place of the target.
+func writeFile(dir, filename string, content []byte) error {
+	// create target folder if it does not exists yet
+	if err := os.MkdirAll(dir, 0700); err != nil {
```
https://kubernetes.slack.com/archives/C0BP8PW9G/p1741888403204129 looks like 0700 should be ok

stlaz commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+	if f.intentCounters[image] > 0 {
+		return
+	}
+	delete(f.intentCounters, image)
```
Will follow 2. Also noticed there was no unit test for the behavior of multiple in-flight pulls, will add that too.

stlaz commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,472 @@
+		return
+	}
+
+	f.decrementImagePullIntent(image)
```
The behavior is desirable, explained in a comment in the code :+1:

liggitt commented on pkg/kubelet/images/pullmanager/locks.go:
```diff
@@ -0,0 +1,67 @@
+
+// NewStripedLockSet creates a StripedLockSet with `size` number of locks to be
+// used for locking context based on string keys.
+// The size will be normalized to stay in the <1, 31> interval.
```
min of 1 makes sense, why a max of 31?

stlaz replied:
I followed the original comment which suggested max of 10... Should this be unbounded from top instead?

liggitt replied:
not necessarily, I just wondered what the significance of 31 was and how it was chosen

liggitt commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -286,6 +287,32 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration, featur
+
+	if localFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) {
+		switch kc.ImagePullCredentialsVerificationPolicy {
+		case "", "NeverVerify",
+			"NeverVerifyPreloadedImages",
+			"NeverVerifyAllowlistedImages",
+			"AlwaysVerify":
```
nit: use the API constants defined for these values here and in the default case error message

liggitt replied:
rather than tolerating "" here, conditionally default "" to NeverVerifyPreloadedImages in SetDefaults_KubeletConfiguration if the feature is enabled

```go
	if localFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) {
		if obj.ImagePullCredentialsVerificationPolicy == "" {
			obj.ImagePullCredentialsVerificationPolicy = NeverVerifyPreloadedImages
		}
	}
```

liggitt commented on staging/src/k8s.io/kubelet/config/v1beta1/types.go:
```diff
@@ -210,6 +211,28 @@ type KubeletConfiguration struct {
+	//   - AlwaysVerify
+	//       - all images require credential reverification
+	// +optional
+	ImagePullCredentialsVerificationPolicy v1alphav1.ImagePullCredentialsVerificationPolicy `json:"imagePullCredentialsVerificationPolicy,omitempty"`
```
`v1alphav1.ImagePullCredentialsVerificationPolicy` should be defined in this package, not v1alpha1

liggitt commented on staging/src/k8s.io/kubelet/config/v1alpha1/types.go:
```diff
@@ -20,6 +20,25 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// ImagePullCredentialsVerificationPolicy is an enum for the policy that is enforced
+// when pod is requesting an image that appears on the system
+type ImagePullCredentialsVerificationPolicy string
+
+const (
+	// NeverVerify will never require credential verification for images that
+	// already exist on the node
+	NeverVerify ImagePullCredentialsVerificationPolicy = "NeverVerify"
+	// NeverVerifyPreloadedImages does not require credential verification for images
+	// pulled outside the kubelet process
+	NeverVerifyPreloadedImages ImagePullCredentialsVerificationPolicy = "NeverVerifyPreloadedImages"
+	// NeverVerifyAllowlistedImages does not require credential verification for
+	// a list of images that were pulled outside the kubelet process
+	NeverVerifyAllowlistedImages ImagePullCredentialsVerificationPolicy = "NeverVerifyAllowlistedImages"
+	// AlwaysVerify requires credential verification for accessing any image on the
+	// node irregardless how it was pulled
+	AlwaysVerify ImagePullCredentialsVerificationPolicy = "AlwaysVerify"
+)
```
move these to the v1beta1 package

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_policies.go:
```diff
@@ -0,0 +1,171 @@
+	switch policy {
+	case kubeletconfiginternal.NeverVerify:
+		return NeverVerifyImagePullPolicy(), nil
+	case "", kubeletconfiginternal.NeverVerifyPreloadedImages:
```
```suggestion
	case kubeletconfiginternal.NeverVerifyPreloadedImages:
```

liggitt commented on pkg/kubelet/images/pullmanager/fs_pullrecords.go:
```diff
@@ -0,0 +1,301 @@
+		decoder: kubeletConfigDecoder,
+	}
+
+	if err := os.MkdirAll(accessor.pullingDir, 0700); err != nil {
+		return nil, err
+	}
+
+	if err := os.MkdirAll(accessor.pulledDir, 0700); err != nil {
+		return nil, err
+	}
```
did you ask about this and get a confirmation 0700 was correct?

liggitt replied:
oh, just saw the other thread 👍

liggitt commented on pkg/kubelet/images/pullmanager/image_pull_manager.go:
```diff
@@ -0,0 +1,553 @@
+	}
+	intentNum, ok := intentNumAny.(int32)
+	if !ok {
+		panic(fmt.Sprintf("expected the intentCounters sync map to only contain int32 values, got %T", intentNum))
```
```suggestion
		panic(fmt.Sprintf("expected the intentCounters sync map to only contain int32 values, got %T", intentNumAny))
```

liggitt commented:
/lgtm
/approve
for API and implementation review

Go ahead and add a release note to the description as well

/retest
/hold to let relevant e2e tests run, and in case any node reviewers want to do a pass on the final changes

SergeyKanzhelev commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -116,15 +131,32 @@ func (m *imageManager) logIt(objRef *v1.ObjectReference, eventtype, event, prefi
+//
+// Called in 2 scenarios:
+//  1. image is not present with `imagePullPolicy: Never“
+//  2. image is present but cannot bne accessed with the presented set of credentials
```
```suggestion
//  2. image is present but cannot be accessed with the presented set of credentials
```

SergeyKanzhelev commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -116,15 +131,32 @@ func (m *imageManager) logIt(objRef *v1.ObjectReference, eventtype, event, prefi
+// We don't want to reveal the presence of an image if it cannot be accessed, hence we
+// want the same behavior in both the above scenarios.
+//
+// TODO: it feels like error handling refactoring might remove the need for this
```
let's not add TODO that is too broad

liggitt commented:
@jpbetz do you know if something just merged with declarative validation that would break the alpha-beta e2e? https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128152/pull-kubernetes-e2e-kind-alpha-beta-features/1900570126353174528 is failing with things like `{ failed [FAILED] Failed to update scale subresource: ReplicationController "e2e-rc-v9xzq" is invalid: <nil>: Internal error: no validation found for *v1.ReplicationController, subresources: [scale]: ReplicationController "e2e-rc-v9xzq" is invalid: <nil>: Internal error: no validation found for *v1.ReplicationController, subresources: [scale]`

SergeyKanzhelev commented on pkg/kubelet/images/pullmanager/interfaces.go:
```diff
@@ -0,0 +1,110 @@
+	// RecordPullIntent() must match exactly one call of RecordImagePulled()/RecordImagePullFailed().
+	//
+	// `image` is the content of the pod's container `image` field.
+	RecordPullIntent(image string) error
```
when we pull, we pass annotations and a handler to the Pull CRI API. This cache seems to assume that the name of the image is a sufficient identification. Will this cause any issues if runtime uses different pulling depending on some attribute? Was it discussed as part of a KEP or we have a new assumption we need to document

stlaz replied:
Annotations were discussed, though not exactly from this perspective.

I would expect that even though annotations and a handler are being passed, the only authentication input specific to an image being pulled would still remain to be docker credentials either from a secret or a node-global keyring (or, an SA token with one of the new keps), and the annotations and handler would just prescribe how these credentials should be used.

My assumption may just as well be wrong ofc, I do not know the CRI API too well, in which case it would make sense to document how the annotations and handler matter.

SergeyKanzhelev replied:
this assumption will need to be documented as a CRI API assumption.

stlaz replied:
@dshebib when you're documenting this, I think you can mention that kube admission should be used in these cases. It's probably not different from the current state but worth commenting on as from the feature description, one might expect that something was going to change in this regard.

mikebrow replied:
@stlaz containerd basically in order to properly pull and unpack a container image the containerd runtime needs to know the runtime handler [11:43](https://kubernetes.slack.com/archives/C0BP8PW9G/p1742316188051309?thread_ts=1741967504.925089&cid=C0BP8PW9G)
annotations was the original method for passing through said information
we added an additional field for runtime handler to the pull image request struct
the reason containerd needs this is each runtime handler can be configured with a different fs snapshotter

https://github.com/containerd/containerd/blob/main/internal/cri/server/images/image_pull.go#L187 There and about to see us start the pull and unpack with the handler’s configured snapshotter … this goes for 2.0+
 
and here for the 1.7 annotations impl: https://github.com/containerd/containerd/blob/release/1.7/pkg/cri/sbserver/image_pull.go#L789

jpbetz commented:
> @jpbetz do you know if something just merged with declarative validation that would break the alpha-beta e2e? [prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128152/pull-kubernetes-e2e-kind-alpha-beta-features/1900570126353174528](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128152/pull-kubernetes-e2e-kind-alpha-beta-features/1900570126353174528) is failing with things like `{ failed [FAILED] Failed to update scale subresource: ReplicationController "e2e-rc-v9xzq" is invalid: <nil>: Internal error: no validation found for *v1.ReplicationController, subresources: [scale]: ReplicationController "e2e-rc-v9xzq" is invalid: <nil>: Internal error: no validation found for *v1.ReplicationController, subresources: [scale]`
> 
> edit: looks like [#130725 (comment)](https://github.com/kubernetes/kubernetes/pull/130725#issuecomment-2725165727)

Yes, https://github.com/kubernetes/kubernetes/pull/130819 fixes

SergeyKanzhelev commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -285,11 +287,37 @@ func NewKubeGenericRuntimeManager(
+			return nil, nil, err
+		}
+
+		fsRecordAccessor, err := imagepullmanager.NewFSPullRecordsAccessor(rootDirectory)
```
the PR for docs will be needed to update this: https://kubernetes.io/docs/reference/node/kubelet-files/

stlaz replied:
the docs PR for the feature can be found here: https://github.com/kubernetes/website/pull/49886

SergeyKanzhelev replied:
Added a note to that PR to update the kubelet-files

SergeyKanzhelev commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -769,3 +788,93 @@ type CrashLoopBackOffConfig struct {
+	// from the CRI.
+	// The filename is a SHA-256 hash of this value. This is to avoid filename-unsafe
+	// characters like ':' and '/'.
+	ImageRef string
```
What if runtime start identifying the image by a different tag after some retagging? This is potential risk we should accept and potentially document. I do not think we have a strict requirement for the runtime how the image ref is defined.

liggitt replied:
This handles that properly, the image -> imageRef resolution is done via the container runtime every time, then the image ref is used to check what credentials are considered good. If the imageRef changes, there won't be any credentials considered valid and it'll fall back to attempting pull

SergeyKanzhelev replied:
Yes, I understand that things will not break. My suggestion is to highlight the fact in docs that kubelet do not guarantee this check if image was re-tagged

liggitt replied:
I'm not really sure what you're suggesting we highlight... if the image is tagged remotely and you have image pull policy other than Always, the kubelet might not notice? That's not new or unique to this feature

stlaz commented:
I fixed the typo in the comment + removed the TODO, PTAL if the latest changes are good :slightly_smiling_face:

stlaz commented:
The failures in pull-kubernetes-node-e2e-containerd-alpha-features don't appear related and are consistent with the Job History outside this PR.

wendy-ha18 commented:
Hello @stlaz @pacoxu

Appreciate all of your efforts with this PR!

A gentle reminder that the code freeze has started [02:00 UTC Friday 21st March 2025](https://everytimezone.com/s/2c5e9275) . Please make sure this PR has both `lgtm` and `approved` labels ASAP, and file an [Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) if you haven't done it yet.
Thanks!

liggitt commented:
/hold cancel
